util/types/time.go

```diff
@@ -486,12 +486,22 @@ func (t *Time) Sub(t1 *Time) Duration {
 	}
 }
 
-// Add adds d to t, returns a duration value.
-// Note that add should not be done on different time types.
-func (t *Time) Add(d Duration) Duration {
-	d.Duration = gotime.Duration(-int64(d.Duration))
-	t2, _ := d.ConvertToTime(t.Type)
-	return t.Sub(&t2)
+// Add adds d to t, returns the result time value.
+func (t *Time) Add(d Duration) (Time, error) {
+	sign, hh, mm, ss, micro := splitDuration(d.Duration)
+	seconds, microseconds, _ := calcTimeDiff(t.Time, FromDate(0, 0, 0, hh, mm, ss, micro), -sign)
+	days := seconds / secondsIn24Hour
+	year, month, day := getDateFromDaynr(uint(days))
+	var tm mysqlTime
+	tm.year, tm.month, tm.day = uint16(year), uint8(month), uint8(day)
+	calcTimeFromSec(&tm, seconds%secondsIn24Hour, microseconds)
+	ret := Time{
+		Time:     tm,
+		Type:     t.Type,
```

Comment:
If t.Type == mysql.TypeDate
we still set the type of result to mysql.TypeDate?

Reply:
Yes.
Do you mean hh:mm:ss should be truncated?